### PR TITLE
Add FXIOS-6502 [v115] middleware and StoreSubscriber component

### PR DESCRIPTION
**Pull request by yoanarios**
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-6502)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/14597)

### Description
- Add Middleware component
- Add StoreSubscriber 

### Pull requests checks where applicable
- [x] Fill in the three TODOs above (tickets number and description of your work)
- [x] The PR name follows our [naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Accessibility implemented and tested (minimum Dynamic Text and VoiceOver)
- [ ] Unit tests written and passing
- [x] Documentation / comments for complex code and public methods

### CODE REVIEW

File: BrowserKit/Sources/Redux/Middleware.swift [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+
+/// Redux `Middleware` provides a third-party extension point between dispatching an `Action`, and the moment it reaches the `Reducer`
+/// Middleware  produces side effects or uses dependencies and
```
Comment (lmarceau) [inline BrowserKit/Sources/Redux/Middleware.swift:8]:
```suggestion
/// Middleware produces side effects or uses dependencies and
```

File: BrowserKit/Sources/Redux/StoreSubscriber.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+
+public protocol AnyStoreSubscriber {
+    func newState(state: Any)
+}
+
+/// Subscribers listen for state changes.
+///  While all reducers will be called when a action gets dispatched, the subscriber will only be called when an actual state change happens.
```
Comment (lmarceau) [inline BrowserKit/Sources/Redux/StoreSubscriber.swift:12]:
```suggestion
/// While all reducers will be called when a action gets dispatched, the subscriber will only be called when an actual state change happens.
```

**Review by lmarceau — COMMENTED** [review]
nits!

**Comment by yoanarios** [comment]
Green build before nits fix [here](https://app.bitrise.io/app/6c06d3a40422d10f/pipelines/53d5ad1b-9622-4861-9903-f81be2c48814)

**Review by OrlaM — APPROVED** [review]

**Pull request merged**